examples/auction/src/lib.rs

```diff
@@ -1,212 +1,197 @@
 fn auction_bid<S: HasStateApi>(
     ctx: &impl HasReceiveContext,
-    host: &mut impl HasHost<State<S>, StateApiType = S>,
+    host: &mut impl HasHost<State, StateApiType = S>,
     amount: Amount,
 ) -> Result<(), BidError> {
+    let balance = host.self_balance();
     let state = host.state_mut();
-    ensure!(
-        state.viewable_state.auction_state == AuctionState::NotSoldYet,
-        BidError::AuctionFinalized
-    );
+    // Ensure the auction has not been finalized yet
+    ensure!(state.auction_state == AuctionState::NotSoldYet, BidError::AuctionAlreadyFinalized);
 
     let slot_time = ctx.metadata().slot_time();
-    ensure!(
-        slot_time <= state.viewable_state.expiry,
-        BidError::BidsOverWaitingForAuctionFinalization
-    );
+    // Ensure the auction has not ended yet
+    ensure!(slot_time <= state.end, BidError::BidTooLate);
 
+    // Ensure that only accounts can place a bid
     let sender_address = match ctx.sender() {
-        Address::Contract(_) => bail!(BidError::ContractSender),
+        Address::Contract(_) => bail!(BidError::OnlyAccount),
         Address::Account(account_address) => account_address,
     };
-    let mut bid_to_update = state.bids.entry(sender_address).or_insert(Amount::zero());
-    *bid_to_update += amount;
-    // Ensure that the new bid exceeds the highest bid so far
-    ensure!(*bid_to_update > state.viewable_state.highest_bid, BidError::BidTooLow);
-
-    state.viewable_state.highest_bid = *bid_to_update;
 
+    // Ensure that the new bid exceeds the highest bid so far
+    ensure!(amount > balance, BidError::BidTooLow);
+
+    if let Some(account_address) = state.highest_bidder.replace(sender_address) {
+        // Refunding old highest bidder;
+        // This transfer (given enough NRG of course) always succeeds because the
+        // `account_address` exists since it was recorded when it placed a bid.
+        // If an `account_address` exists, and the contract has the funds then the
+        // transfer will always succeed.
+        // Please consider using a pull-over-push pattern when expanding this smart
+        // contract to allow smart contract instances to participate in the auction as
+        // well. https://consensys.github.io/smart-contract-best-practices/attacks/denial-of-service/
+        let _ = host.invoke_transfer(&account_address, balance);
+    }
     Ok(())
 }
 
-/// View function that returns the contents of the state except the map of
-/// individual bids.
-#[receive(contract = "auction", name = "view", return_value = "ViewableState")]
+/// View function that returns the content of the state
+#[receive(contract = "auction", name = "view", return_value = "State")]
 fn view<S: HasStateApi>(
     _ctx: &impl HasReceiveContext,
-    host: &impl HasHost<State<S>, StateApiType = S>,
-) -> ReceiveResult<ViewableState> {
-    Ok(host.state().viewable_state.clone())
+    host: &impl HasHost<State, StateApiType = S>,
+) -> ReceiveResult<State> {
```

Comment:
In principle returning `&State` here should work (although you might need a few lifetime annotations) and would improve performance. I'm not sure we have an implementation of Serial for `&A` though. If not, it would be useful to have it in cases like this. So an implementation like

```rust
impl <'a, A: Serial> Serial for &'a A {

}
```